lib/helpers.bash

```diff
@@ -849,3 +849,22 @@ then
     fi
   }
 fi
+
+function _bash_it_find_in_ancestor() (
+	# We're deliberately using a subshell for this entire function for simplicity.
+	# By using a subshell, we can use ${PWD} without special handling, and can
+	#  just `cd ..` to move up the directory heirarchy.
+	# Let the shell do the work.
+	local common_ancestor="${PWD}" kin
+	while [[ "${PWD:-/}" != '/' ]]; do
+		for kin in "$@"; do
+			if [[ -r "${PWD}/${kin}" ]]; then
+				common_ancestor="${PWD}"
+				echo "${common_ancestor}"
+				return 0
+			fi
+		done
+		command cd .. || return 2
+	done
+	return 1
+)
```

Comment:
I'll add properly descriptive comment; good call. I did specifically say it invokes it's own subshell (caller doesn't need to subshell for our needs). Return codes match _Bash_ built-ins: 0 = success, 1 = not success, 2 = wrong usage.

Very short version: the purpose is to figure out if we're inside a git repo by searching for an ancestor which has a `.git` folder.

Longer version: lots of tools store special/configuration file in *any* *parent* *of* the current folder, including `.editorconfig`, `.shellcheckrc`, `.git`, `.gradle`, &c. &c. This function's purpose is to just figure out "true or false, we are inside of something governed by X". Specifically the `gradle` plugin searches for one of two different files, either of which will return success; so I made sure to handle that use case as well. The caller may or may not need the path to the config file; this kind of check is useful to determine if we should call `git` or not, rather than to actually do something with the `.git` folder itself. E.g., if we're not inside a git repo, then don't even bother calling `git`.

`common_ancestor` is useless in this version; it's a left-over vestige from an earlier implementation. I'll remove it.

Not checking for `/.git` or `/Makefile` is intentional. If we're in root, then we're no longer looking for "are we in a `gradle` environment", so there's no sense checking for it. I suppose one could search for `/etc`, but I feel like that would be weird to do.

👍 `printf`

---

An express goal of my writing this function is to be straightforward to read and straightforward to run. It looks like it climbs up the directory tree, which is exactly what it does. I specifically don't want to get strings returned by an external tool that need to be interpreted for newlines, spaces, &c. I don't want to have to prepare arguments for a utility, &c. _Bash_ knows about current working directory and how to change directory and I want to let it do what it's good at!

Reply:
ok, I've just pushed an updated version

Reply:
Just to be clear, this function creates it's own subshell by providing the command list within `(``)` instead of `{``}`, so the user of the function can just do what they want and not worry about what we're doing. This was an express design goal to avoid special usage requirements.